randomizer.py

```diff
@@ -990,63 +1338,158 @@ def nearest_cluster(self):
             if self.nearest_exit in c.exits:
                 return c
 
+    def replace_item(self, old_item, new_item):
+        assert(self.chest_contents == old_item)
+        self.tpt.argument = new_item.index
+
     @classmethod
     def mutate_all(cls):
         cls.class_reseed("mut")
-        objs = list(cls.every)
-        random.shuffle(objs)
-        for o in cls.every:
-            if hasattr(o, "mutated") and o.mutated:
-                continue
-            o.reseed(salt="mut")
-            o.mutate()
-            o.mutate_bits()
-            o.mutated = True
+        if 'a' not in get_flags():
+            for o in cls.every:
+                if hasattr(o, "mutated") and o.mutated:
+                    continue
+                o.reseed(salt="mut")
+                o.mutate()
+                o.mutate_bits()
+                o.mutated = True
+            return
+            
+        # Ancient Cave
+        # 0) Set non-in-cave chests to empty, for spoiler clarity
+        # Also set chests that are unreachable to be empty
+        inaccessible_chests = [o for o in cls.every if o.is_chest and o.cave_rank is None and (not hasattr(o, "mutated") or not o.mutated)] 
+        inaccessible_chests += [
+            cls.get(182), cls.get(135),                 # Dungeon man
+            ]
+        for chest in inaccessible_chests:
+            chest.tpt.argument = 0x100
+            chest.mutated = True
+
+        # 1) Place skip-granting items early in the cave
+        early_items_index = [
+            0x7d,   # Backstage pass
+            0xa6,   # King banana
+            0xb8,   # Pencil eraser
+            0xd2,   # Eraser eraser
+            0xfd,   # Carrot key
+        ]
+        early_chests = sorted(cls.unassigned_chests, key=lambda c: c.cave_rank)
+        early_chests = early_chests[:len(early_chests)/3]
```

Comment:
I don't mind if you want to weight these items toward the early parts of the cave, but I'd recommend against removing them from the latter parts altogether. In my view, finding a key that gives you a potential reason to backtrack to an earlier map to use it is a many times more interesting gameplay mechanic than reaching a locked door with the key already in your possession.

I'd recommend that they have the potential to appear anywhere in the dungeon, with diminishing likelihood the deeper into the dungeon you go.

Reply:
I see your point, but having watch a few playthroughs, the value of taking a skip decreases rapidly the deeper you get. After all, at halfway through, there’s a 50/50 shot the skip would take you backwards, and they do take a not-insigificant amount of time to investigate to find out where you are and if you’re going forwards or backwards—you basically have to go until you find a shiny spot or a previously found location. 

At 75% in I think most people going for speed wouldn’t even take a skip if one showed up free right in front of them, let alone backtrack for one. I do see your point that they may be too frontloaded as is though. Will prolly adjust this next version, maybe to first 50 or 60 percent of locations? Thoughts?

Reply:
I figure there's value in simply being able to shortcut back to a previous area. If you're lucky and you've been paying attention, the shortcut might take you near an area with a shop or inn or phone, and that's useful even on the last few maps. Edge cases like that have value because they are rare; I wouldn't want to restrict the number of possible layouts of the dungeon and make it predictable to cater to speedrunners. Generally my design philosophy for randomizers is to make them such that anything is possible, but I agree that probabilities should be weighted heavily in favor of a balanced playthrough.

I'm planning to activate the custom randomness settings for this randomizer. If you like, we could link this feature to that, which would allow the user to roughly decide how unpredictable they want these items to be.

Reply:
Yeah, that's still on the agenda. Sorry it's taken so long, I've been busy with IRL stuff.